src/core/bitop.d

```diff
@@ -194,10 +256,149 @@ version (DigitalMars) version (AnyX86) @system // not pure
     uint outpl(uint port_address, uint value);
 }
 
+
+/**
+ *  Calculates the number of set bits in an integer.
+ */
+int popcnt(uint x) pure
+{
+    // Select the fastest method depending on the compiler and CPU architecture
+    version(LDC)
+    {
+        return _popcnt(x);
+    }
+    else
+    {
+        version(DigitalMars)
+        {
+            static if(is(typeof(_popcnt(uint.max))))
+            {
+                import core.cpuid;
+                if(hasPopcnt)
+                    return _popcnt(x);
+            }
+        }
+
+        return soft_popcnt!uint(x);
+    }
+}
+
+unittest
+{
+    assert( popcnt( 0 ) == 0 );
+    assert( popcnt( 7 ) == 3 );
+    assert( popcnt( 0xAA )== 4 );
+    assert( popcnt( 0x8421_1248 ) == 8 );
+    assert( popcnt( 0xFFFF_FFFF ) == 32 );
+    assert( popcnt( 0xCCCC_CCCC ) == 16 );
+    assert( popcnt( 0x7777_7777 ) == 24 );
+}
+
+/// ditto
+int popcnt(ulong x) pure
+{
+    // Select the fastest method depending on the compiler and CPU architecture
+    version(LDC)
+    {
+        return _popcnt(x);
+    }
+    else
+    {
+        import core.cpuid;
+
+        static if(size_t.sizeof == uint.sizeof)
+        {
+            const sx = Split64(x);
+            version(DigitalMars)
+            {
+                static if(is(typeof(_popcnt(uint.max))))
+                {
+                    if(hasPopcnt)
+                        return _popcnt(sx.lo) + _popcnt(sx.hi);
+                }
+            }
+
+            return soft_popcnt!uint(sx.lo) + soft_popcnt!uint(sx.hi);
+        }
+        else
+        static if(size_t.sizeof == ulong.sizeof)
+        {
+            version(DigitalMars)
+            {
+                static if(is(typeof(_popcnt(ulong.max))))
+                {
+                    if(hasPopcnt)
+                        return _popcnt(x);
+                }
+            }
+
+            return soft_popcnt!ulong(x);
+        }
+        else
+            static assert(false);
+    }
+}
+
+unittest
+{
+    assert(popcnt(0uL) == 0);
+    assert(popcnt(1uL) == 1);
+    assert(popcnt((1uL << 32) - 1) == 32);
+    assert(popcnt(0x48_65_6C_6C_6F_3F_21_00uL) == 28);
+    assert(popcnt(ulong.max) == 64);
+}
+
+private int soft_popcnt(N)(N x) pure
+    if(is(N == uint) || is(N == ulong))
+{
+    // Avoid branches, and the potential for cache misses which
+    // could be incurred with a table lookup.
+
+    // We need to mask alternate bits to prevent the
+    // sum from overflowing.
+    // add neighbouring bits. Each bit is 0 or 1.
+    enum mask1 = cast(N)0x5555_5555_5555_5555L;
```

Comment:
> Do we really need a enum name for each step in the process?

No. I just thought it was a little easier to read this way. I can change it if you want.